core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -272,6 +272,19 @@ public CommandsResult reroute(final ClusterState clusterState, AllocationCommand
         allocation.ignoreDisable(false);
         // the assumption is that commands will move / act on shards (or fail through exceptions)
         // so, there will always be shard "movements", so no need to check on reroute
+
+        if (retryFailed) {
+            final RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator = routingNodes.unassigned().iterator();
+            while (unassignedIterator.hasNext()) {
+                final ShardRouting shard = unassignedIterator.next();
+                UnassignedInfo unassignedInfo = shard.unassignedInfo();
+                UnassignedInfo newUnassignedInfo = new UnassignedInfo(unassignedInfo.getReason(), unassignedInfo
```

Comment:
I think we can add a resetFailedAllocations method to UnassignedInfo that returns a new copy where the count is reset